engine-src/Game/LambdaHack/Client/UI/SessionUI.hs

```diff
@@ -55,9 +55,9 @@ data SessionUI = SessionUI
   , shistory       :: History       -- ^ history of messages
   , spointer       :: K.PointUI     -- ^ mouse pointer position
   , slastAction    :: Maybe K.KM    -- ^ last pressed key
-  , smacroBuffer   :: Either [K.KM] [K.KM]
+  , smacroBuffer   :: Either Macro [K.KM]
```

Comment:
With all the changes and discussion I'd put macros in the right component of Either, because traditionally the left component is for (potentially) wrong things and the right one is for right things. And the macros, being validated and bound by invariants, are right, as oppose to the recorded things, which are not quite right yet. :)

Reply:
That's all good and right. I just hesitated to bring up the bifunctors and was to lazy to write the whole thing explicitly :D